File: leakcanary-android/src/main/java/com/squareup/leakcanary/internal/HeapAnalyzerService.java
/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.leakcanary.internal;

import android.content.Context;
import android.content.Intent;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import com.squareup.leakcanary.AbstractAnalysisResultService;
import com.squareup.leakcanary.AnalysisResult;
import com.squareup.leakcanary.AnalyzerProgressListener;
import com.squareup.leakcanary.CanaryLog;
import com.squareup.leakcanary.HeapAnalyzer;
import com.squareup.leakcanary.HeapDump;
import com.squareup.leakcanary.R;
import java.io.File;
import java.io.IOException;
import java.util.List;

import static com.squareup.leakcanary.internal.LeakCanaryInternals.setEnabledBlocking;

/**
 * This service runs in a separate process to avoid slowing down the app process or making it run
 * out of memory.
 */
public final class HeapAnalyzerService extends ForegroundService
    implements AnalyzerProgressListener {

  private static final String LISTENER_CLASS_EXTRA = "listener_class_extra";
  private static final String HEAPDUMP_EXTRA = "heapdump_extra";

  public static void runAnalysis(Context context, HeapDump heapDump,
      Class<? extends AbstractAnalysisResultService> listenerServiceClass) {
    setEnabledBlocking(context, HeapAnalyzerService.class, true);
    setEnabledBlocking(context, listenerServiceClass, true);
    Intent intent = new Intent(context, HeapAnalyzerService.class);
    intent.putExtra(LISTENER_CLASS_EXTRA, listenerServiceClass.getName());
    intent.putExtra(HEAPDUMP_EXTRA, heapDump);
    ContextCompat.startForegroundService(context, intent);
  }

  public HeapAnalyzerService() {
    super(HeapAnalyzerService.class.getSimpleName(), R.string.leak_canary_notification_analysing);
  }

  @Override protected void onHandleIntentInForeground(@Nullable Intent intent) {
    if (intent == null) {
      CanaryLog.d("HeapAnalyzerService received a null intent, ignoring.");
      return;
    }
    String listenerClassName = intent.getStringExtra(LISTENER_CLASS_EXTRA);
    HeapDump heapDump = (HeapDump) intent.getSerializableExtra(HEAPDUMP_EXTRA);

    HeapAnalyzer heapAnalyzer =
        new HeapAnalyzer(heapDump.excludedRefs, this, heapDump.reachabilityInspectorClasses);

    List<AnalysisResult> analysisResults =
        heapAnalyzer.checkForLeaks(heapDump.heapDumpFile, heapDump.computeRetainedHeapSize);

    int i = 0;
    for (AnalysisResult result : analysisResults) {
      HeapDump fakeFileHeapDump;
      if (i > 0) {
        // TODO 1 analysis = many leaks, which is currently unsupported by the rest of the pipeline.
        // We temporarily ignore this problem by replacing the heapdump file with a fake file for
        // all heap dumps but the first one.
        File newFile = new File(heapDump.heapDumpFile.getParentFile(),
            i + heapDump.heapDumpFile.getName());
        try {
          boolean created = newFile.createNewFile();
          if (!created) {
            continue;
          }
        } catch (IOException e) {
          continue;
        }
        fakeFileHeapDump = heapDump.buildUpon()
            .heapDumpFile(
                newFile)
            .build();
      } else {
        fakeFileHeapDump = heapDump;
      }
      AbstractAnalysisResultService.sendResultToListener(this, listenerClassName, fakeFileHeapDump,
          result);
      i++;
    }
  }

  @Override public void onProgressUpdate(Step step) {
    int percent = (int) ((100f * step.ordinal()) / Step.values().length);
    CanaryLog.d("Analysis in progress, working on: %s", step.name());
    String lowercase = step.name().replace("_", " ").toLowerCase();
    String message = lowercase.substring(0, 1).toUpperCase() + lowercase.substring(1);
    showForegroundNotification(100, percent, false, message);
  }
}

File: leakcanary-android/src/main/java/leakcanary/internal/HeapAnalyzerService.kt
/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package leakcanary.internal

import android.content.Context
import android.content.Intent
import androidx.core.content.ContextCompat
import com.squareup.leakcanary.R
import leakcanary.AnalyzerProgressListener
import leakcanary.CanaryLog
import leakcanary.HeapAnalyzer
import leakcanary.HeapDump

/**
 * This service runs in a separate process to avoid slowing down the app process or making it run
 * out of memory.
 */
internal class HeapAnalyzerService : ForegroundService(
    HeapAnalyzerService::class.java.simpleName, R.string.leak_canary_notification_analysing
), AnalyzerProgressListener {

  override fun onHandleIntentInForeground(intent: Intent?) {
    if (CanaryLog.logger == null) {
      CanaryLog.logger = DefaultCanaryLog()
    }

    if (intent == null) {
      CanaryLog.d("HeapAnalyzerService received a null intent, ignoring.")
      return
    }
    val heapDump = intent.getSerializableExtra(HEAPDUMP_EXTRA) as HeapDump

    val heapAnalysis =
      if (heapDump.useExperimentalHeapParser) {
        leakcanary.experimental.HeapAnalyzer(this)
            .checkForLeaks(heapDump)
      } else HeapAnalyzer(this).checkForLeaks(heapDump)

    AnalysisResultService.sendResult(this, heapAnalysis)
  }

  override fun onProgressUpdate(step: AnalyzerProgressListener.Step) {
    val percent = (100f * step.ordinal / AnalyzerProgressListener.Step.values().size).toInt()
    CanaryLog.d("Analysis in progress, working on: %s", step.name)
    val lowercase = step.name.replace("_", " ")
        .toLowerCase()
    val message = lowercase.substring(0, 1).toUpperCase() + lowercase.substring(1)
    showForegroundNotification(100, percent, false, message)
  }

  companion object {

    private const val HEAPDUMP_EXTRA = "heapdump_extra"

    fun runAnalysis(
      context: Context,
      heapDump: HeapDump
    ) {
      val intent = Intent(context, HeapAnalyzerService::class.java)
      intent.putExtra(HEAPDUMP_EXTRA, heapDump)
      ContextCompat.startForegroundService(context, intent)
    }
  }
}
